File: heat.cpp
/*! MPI TUTORIAL - solve heat equation in 1d --> C++ VERSION
 -
 - A. Yeates, March 2017
 */
#include <mpi.h> // <-- note this
#include <math.h>
#include <iostream>
#include <fstream>
using namespace std;

int main(){
    static const int nx=2048;
    static const double k=1.0, tMax=0.1;
    int i;
    
    // (A) Initialize MPI
    // ==================
    
    MPI_Init(NULL, NULL);
    
    // - Get number of MPI processes and global rank:
    int nProcs, myRank;
    MPI_Comm_size(MPI_COMM_WORLD, &nProcs);
    MPI_Comm_rank(MPI_COMM_WORLD, &myRank);
    /*if(myRank==0) printf("%d processes\n", nProcs);
    printf("Hello from process %d\n", myRank);*/
    
    // - Set up Cartesian communicator:
    int dims[1], periodic[1];
    dims[0] = nProcs;
    periodic[0] = 0;
    MPI_Comm comm;
    MPI_Cart_create(MPI_COMM_WORLD, 1, dims, periodic, 1, &comm);
    MPI_Comm_rank(comm, &myRank);
    /*printf("Hello from process %d\n", myRank);*/
    
    // - Get ranks of neighbours:
    int prvRank, nxtRank;
    MPI_Cart_shift(comm, 0, 1, &prvRank, &nxtRank);
    /*printf("%d has neighbours %d and %d\n", myRank, prvRank, nxtRank);*/
    
    // (B) Initial condition
    // =====================
    
    // - Check that grid divides evenly between processes:
    if((nx % nProcs) > 0){
        if(myRank==0){
            printf("%d grid points don't divide evenly into %d processes\n", nx, nProcs);
            return 0;
            }
        }
    int nxLocal = nx/nProcs;
    double dx = 2.0/(double(nx) - 1.0);
    /*printf("%g\n",dx);*/
    
    // - One process defines coordinate array (global):
    double* xGlob; // --> need to define pointer so others can call MPI_Scatter.
    if(myRank==0){
        xGlob = new double[nx];
        for(i=0; i<nx; i++){
            xGlob[i] = -1.0 + double(i)*dx;
            }
        }
    
    // - Distribute chunks of coordinate array to each process:
    double x[nxLocal];
    MPI_Scatter(xGlob, nxLocal, MPI_DOUBLE, x, nxLocal, MPI_DOUBLE, 0, comm);
    /*printf("%d %g %g\n",myRank,x[0],x[nxLocal-1]);*/
    
    // - Define initial condition function (locally for each process):
    double g[nxLocal];
    for(i=0; i<nxLocal; i++){
        g[i] = exp(-10.0*x[i]*x[i]);
        }
    
    // - Set timestep (for stability):
    double dt = 0.2*dx*dx/k;
    double mu = k*dt/(dx*dx);
    
    // - Initialize u (locally, including ghost cells):
    double u[nxLocal+2];
    double uNew[nxLocal];
    u[0] = 0.0;
    u[nxLocal+1] = 0.0;
    for(i=1; i<=nxLocal; i++){
        u[i] = g[i-1];
        }

    // (C) Main loop
    // =============

    double start, finish;
    start=MPI_Wtime();
    
    double t = 0.0;
    double uav, uavGlob;
    while(t < tMax){
        // - Enforce global boundary conditions:
        if(prvRank==MPI_PROC_NULL) u[0] = u[2];
        if(nxtRank==MPI_PROC_NULL) u[nxLocal+1] = u[nxLocal-1];
        
        // - Communicate ghost values of u:
        // -- send to left:
        if(prvRank!=MPI_PROC_NULL) MPI_Send(&u[1], 1, MPI_DOUBLE, prvRank, 13, comm);
        if(nxtRank!=MPI_PROC_NULL) MPI_Recv(&u[nxLocal+1], 1, MPI_DOUBLE, nxtRank, 13, comm, MPI_STATUS_IGNORE);
        // -- send to right:
        if(nxtRank!=MPI_PROC_NULL) MPI_Send(&u[nxLocal], 1, MPI_DOUBLE, nxtRank, 13, comm);
        if(prvRank!=MPI_PROC_NULL) MPI_Recv(&u[0], 1, MPI_DOUBLE, prvRank, 13, comm, MPI_STATUS_IGNORE);
        
        // - Update u:
        for(i=1; i<=nxLocal; i++){
            uNew[i-1] = u[i] + mu*(u[i-1] - 2.0*u[i] + u[i+1]);
            }
         for(i=1; i<=nxLocal; i++){
            u[i] = uNew[i-1];
            }
        
        /*
        // - Compute average of u:
        uav = 0.0;
        for(i=1; i<=nxLocal; i++){
            uav = uav + u[i];
            }
        MPI_Allreduce(&uav, &uavGlob, 1, MPI_DOUBLE, MPI_SUM, comm);
        uavGlob = uavGlob/double(nx);
        if(myRank==0) printf("%f %f\n",t, uavGlob);
        */

        // - Update t:
        t += dt;
        }
    
    finish=MPI_Wtime();
    printf("Parallel Elapsed time: %f seconds\n", finish-start);
    
    // (D) Output results
    // ==================
    
    // - Collect results in control process:
    double uInterior[nxLocal];
    for(i=1; i<=nxLocal; i++){
        uInterior[i-1] = u[i];
        }
    double* uGlob;
    if(myRank==0) uGlob = new double[nx];
    MPI_Gather(uInterior, nxLocal, MPI_DOUBLE, uGlob, nxLocal, MPI_DOUBLE, 0, comm);
    
    // - Save to file:
    if(myRank==0){
        ofstream foo;
        foo.open("u.dat", ios::binary | ios::out);
        foo.write((char *)&nx, sizeof(int));
        for(i=0; i<nx; i++){
            foo.write((char *)&xGlob[i], sizeof(xGlob[i]));
            }
        for(i=0; i<nx; i++){
            foo.write((char *)&uGlob[i], sizeof(uGlob[i]));
            }
        foo.close();
        }

    MPI_Finalize();
    }

File: heat.f90
! MPI TUTORIAL - solve heat equation in 1d --> FORTRAN 90 VERSION
!
! - A. Yeates, March 2017
!
!
program heat
    implicit none
    integer, parameter :: d=kind(0.d0)

    include 'mpif.h'  ! <-- note this

    integer, parameter :: nx=2048
    real(d), parameter :: k = 1.0_d
    real(d), parameter :: tMax = 0.1_d
    real(d) :: dx, dt, t, mu, uav, uavGlob
    integer :: ierr, nProcs, myRank, comm, nxtRank, prvRank, nxLocal, i
    real(d), allocatable :: x(:), g(:), u(:), xGlob(:), uGlob(:)
    real(d) :: start, finish

    ! (A) Initialize MPI
    ! ==================

    call mpi_init(ierr)

    ! - Get number of MPI processes and global rank:
    call mpi_comm_size(MPI_COMM_WORLD, nProcs, ierr)
    call mpi_comm_rank(MPI_COMM_WORLD, myRank, ierr)
!    if (myRank==0) print*,nprocs,' processes'
!    print*,'Hello from process ',myRank

    ! - Set up Cartesian communicator:
    call mpi_cart_create(MPI_COMM_WORLD, 1, nProcs, 0, 1, comm, ierr)
    call mpi_comm_rank(comm, myRank, ierr)

    ! Get ranks of neighbours:
    call mpi_cart_shift(comm, 0, 1, prvRank, nxtRank, ierr)
!    print*,myRank,' has neighbours ',prvRank,' and ',nxtrank
!    print*,'MPI_PROC_NULL=',MPI_PROC_NULL

    ! (B) Initial condition
    ! =====================

    ! - Check that grid divides evenly between processes:
    if (mod(nx, nProcs)>0) then
        if (myRank==0) print*,nx," grid points don't divide evenly among",nProcs,"processors"
        stop
    end if
    nxLocal = nx/nProcs
    dx = 2.0_d/(real(nx,d) - 1.0_d)
!    print*,dx

    ! - One process defines coordinate array (global):
    if (myRank==0) then
        allocate(xGlob(1:nx))
        xGlob = (/ (-1.0_d + real(i)*dx, i=0,nx-1) /)
    end if

    ! - distribute chunks of coordinate array to each process:
    allocate(x(1:nxLocal))
    call mpi_scatter(xGlob, nxLocal, MPI_DOUBLE_PRECISION, x, nxLocal, MPI_DOUBLE_PRECISION, 0, comm, ierr)
!    print*,myRank,x

    ! - define initial condition function (locally for each process):
    allocate(g(1:nxLocal))
    g = exp(-10.0_d*x**2)
!    print*,g

    ! - Set timestep (for stability):
    dt = 0.2_d*dx**2/k
    mu = k*dt/(dx**2)

    ! -Initialize u (locally, including ghost cells):
    allocate(u(0:nxLocal+1))
    u = 0.0_d
    u(1:nxLocal) = g

    ! (C) Main loop
    ! =============

    start=MPI_Wtime();

    t = 0.0_d
    do while (t < tMax)
        ! - Enforce global boundary conditions:
        if (prvRank==MPI_PROC_NULL) u(0) = u(2)
        if (nxtRank==MPI_PROC_NULL) u(nxLocal+1) = u(nxLocal-1)

        ! - Communicate ghost values of u:
        ! -- send to left:
        if (prvRank/=MPI_PROC_NULL) call mpi_send(u(1), 1, MPI_DOUBLE_PRECISION, prvRank, 13, comm, ierr)
        if (nxtRank/=MPI_PROC_NULL) call mpi_recv(u(nxLocal+1), 1, MPI_DOUBLE_PRECISION, nxtRank, 13, comm, MPI_STATUS_IGNORE, ierr)
        ! -- send to right:
        if (nxtRank/=MPI_PROC_NULL) call mpi_send(u(nxLocal), 1, MPI_DOUBLE_PRECISION, nxtRank, 13, comm, ierr)
        if (prvRank/=MPI_PROC_NULL) call mpi_recv(u(0), 1, MPI_DOUBLE_PRECISION, prvRank, 13, comm,  MPI_STATUS_IGNORE, ierr)

        ! - Update u:
        u(1:nxLocal) = u(1:nxLocal) + mu*(u(0:nxLocal-1) - 2.0_d*u(1:nxLocal) + u(2:nxLocal+1))

        ! - Compute average of u:
        !uav = sum(u(1:nxLocal))
        !call mpi_allreduce(uav, uavGlob, 1, MPI_DOUBLE_PRECISION, MPI_SUM, comm, ierr)
        !uavGlob = uavGlob/real(nx,d)
        !if (myRank==0) print*,t,uavGlob

        ! - Update t:
        t = t + dt
    end do

    finish=MPI_Wtime();
    print*,'Parallel Elapsed time:',finish-start,' seconds'

    ! (D) Output results
    ! ==================

    ! - Collect results in control process:
    if (myRank==0) allocate(uGlob(1:nx))
    call mpi_gather(u(1:nxLocal), nxLocal, MPI_DOUBLE_PRECISION, uGlob, nxLocal, MPI_DOUBLE_PRECISION, 0, comm, ierr)

    ! - Save to file:
    if (myRank==0) then
        open(unit=1, file='u.dat', action='write', status='replace', form='unformatted', access='stream')
        write(1) nx
        write(1) xGlob
        write(1) uGlob
        close(unit=1)
    end if

    call mpi_finalize(ierr)

end program heat
